drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * drmModeAtomicCommit(). This function internally creates a struct
+ * drm_mode_atomic object that receives all the properties we want to change.
+ * In the end, a single IOCTL will be called using this drm_mode_atomic object:
+ * DRM_IOCTL(fd, DRM_IOCTL_MODE_ATOMIC, &atomic)
+ */
+
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev)
+{
+	drmModeAtomicReq *req;
+	uint32_t blob_id;
+	int ret, flags;
+	struct drm_object *plane = &dev->plane;
+	struct modeset_buf *buf = &dev->bufs[dev->front_buf ^ 1];
+
+	req = drmModeAtomicAlloc();
+
+	/* prepare our connector object to set the CRTC id that
+	 * the connector is using */
+	if (set_connector_property(req, &dev->connector, dev->conn_id, 
+							   "CRTC_ID", dev->crtc_id) < 0)
+		return -1;
+
+	/* Creates a blob property that receives dev->mode value. Then, the CRTC
+	 * object property "MODE_ID" points to the id of the blob. This usually
+	 * happens for properties that are not simple types. In this case, dev->mode
+	 * is a struct. But we could have a property that expects the id of a blob
+	 * that holds an array, for instance.
+	 */
+	if (drmModeCreatePropertyBlob(fd, &dev->mode, sizeof(dev->mode), &blob_id) != 0)
+		return -1;
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "MODE_ID", blob_id) < 0)
+		return -1;
+
+	/* prepare our CRTC object to be set as active */
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "ACTIVE", 1) < 0)
+		return -1;
+
+	/* prepare to set properties of the plane related to the CRTC and the
+	 * framebuffer */
+	set_plane_property(req, plane, dev->plane_id, "FB_ID", buf->fb);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_ID", dev->crtc_id);
+	set_plane_property(req, plane, dev->plane_id, "SRC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+	set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* Set the connector, CRTC and plane properties in a single IOCTL. Here we
+	 * leave it as a blocking call, since we can't start to do page-flips before
+	 * setting everything up.
+	 */
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * Just like modeset_atomic_modeset(), in modeset_atomic_commit() we're going to
+ * atomically set some properties for our objects. The difference is that here
+ * we are setting up just the planes properties. So this is used when we want to
+ * flip a page and not during the modeset setup.
+ */
+
+static int modeset_atomic_commit(int fd, struct modeset_buf *buf, struct modeset_dev *dev)
```

Comment:
There should be no need for this function to exist. Just call `modeset_atomic_modeset()` instead, but without adding `ALLOW_MODESET` to the flags.

Reply:
I've created a new function named `modeset_perform_modeset()` which perform the test-only and the real modeset. Then, the page-flip only commits are being done inside `modeset_draw_out()`.